lib/experimental/class-wp-html-updater.php

```diff
@@ -0,0 +1,556 @@
+
+		return true;
+	}
+
+	public function set_attribute( $name, $value ) {
+		if ( $this->current_tag ) {
+			$escaped_new_value = esc_attr( $value );
+			$updated_attribute = "{$name}=\"{$escaped_new_value}\"";
+
+			$attr = $this->find_current_tag_attribute( $name );
+			if ( $attr ) {
+				// Update an existing attribute.
+				$diff = new WP_HTML_Diff(
+					$attr->get_start_index(),
+					$attr->get_end_index(),
+					$updated_attribute
+				);
+			} else {
+				// Create a new attribute at the tag's name end.
+				$diff = new WP_HTML_Diff(
+					$this->current_tag_name_ends_at,
+					$this->current_tag_name_ends_at,
+					" {$updated_attribute}"
+				);
+			}
+			$this->add_diff( $name, $diff );
+		}
+
+		return $this;
+	}
+
+	public function remove_attribute( $name ) {
+		if ( $this->current_tag ) {
+			$attr = $this->find_current_tag_attribute( $name );
+			if ( $attr ) {
+				$diff = new WP_HTML_Diff(
+					$attr->get_start_index(),
+					$attr->get_end_index(),
+					''
+				);
+				$this->add_diff( $name, $diff );
+			}
+		}
+
+		return $this;
+	}
+
+	private function find_current_tag_attribute( $name ) {
+		if ( array_key_exists( $name, $this->current_tag_attributes ) ) {
+			return $this->current_tag_attributes[ $name ];
+		}
+		do {
+			$attr = $this->parse_next_attribute();
+		} while ( $attr && ! $attr->name_equals( $name ) );
+
+		return $attr;
+	}
+
+	public function add_class( $class_name ) {
+		if ( $this->current_tag ) {
+			$this->current_class_names_set()->add( $class_name );
+		}
+
+		return $this;
+	}
+
+	public function remove_class( $class_name ) {
+		if ( $this->current_tag ) {
+			$this->current_class_names_set()->remove( $class_name );
+		}
+
+		return $this;
+	}
+
+	private function add_diff( $attribute_name, $diff ) {
+		if ( in_array( $attribute_name, $this->current_tag_touched_attributes_names, true ) ) {
+			throw new WP_HTML_Attribute_Already_Modified( "Only one change per tag attribute is supported and the attribute '{$attribute_name}' was already changed on tag {$this->current_tag}." );
+		}
+		$this->current_tag_touched_attributes_names[] = $attribute_name;
+		$this->diffs[]                                = $diff;
+	}
+
+	private function current_class_names_set() {
+		if ( $this->current_tag_class_names === null ) {
+			$class_attr                    = $this->find_current_tag_attribute( 'class' );
+			$class_value                   = $class_attr ? $class_attr->get_value() : '';
+			$this->current_tag_class_names = new WP_HTML_Updater_Classnames_Set( $class_value );
+		}
+
+		return $this->current_tag_class_names;
+	}
+
+	private function match( $regexp ) {
+		$matches = null;
+		$result  = preg_match(
+			$regexp,
+			$this->html,
+			$matches,
+			PREG_OFFSET_CAPTURE,
+			$this->parsed_characters
+		);
+		if ( 1 !== $result ) {
+			throw new WP_HTML_No_Match();
+		}
+
+		return $matches;
+	}
+
+	private function move_parser_to_match( $match ) {
+		$this->parsed_characters = $this->index_of_match( $match );
```

Comment:
side note: we may want to avoid the term "character" since that's so cursed from a human-text / Unicode perspective. "user perceivable character" in Unicode is roughly what "Extended Grapheme Cluster" means, which itself can be costly to count and recognize, whereas what we're doing here is tracking byte offsets as returned by `preg_match()`. Theoretically, if we mess up on the tracking  and advance this we could advance it _inside_ a character where no boundary exists, and corrupt the text stream. We won't likely do that, but I find value in the specificity of these terms to avoid the entire situation since we're inevitably dealing with those text-related terms.

Reply:
Good point! I renamed this to `parsed_bytes`.